### PR TITLE
add JSON-RPC method to call /storeraw on GoQuorum enclave

### PR DESCRIPTION
**Pull request by macfarla**
## PR description
override existing JSON-RPC method: priv_distributeRawTransaction which is a proxy to Tessera's /storeraw - and the factory chooses which implementation to create, depending on whether go-quorum mode is enabled.

/storeraw is required to process GoQuorum private transactions via EthSigner (analogous to priv_distributeRawTransaction for Orion)

## Fixed Issue(s)
refs https://github.com/ConsenSys/protocol-misc/issues/256

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

```
{
    "jsonrpc": "2.0",
    "method": "goquorum_storeRaw",
    "params": [
        "tQEmN0d/xXJZs5OMgl8QVBIyYxu1XubAKehsSYbcOjbxai+QJQpEOs6ghrYAZizLtnM4EJdMyVeVrxO3cA9JJA=="
    ],
    "id": 1
}
```
Gives result:
```
{
    "jsonrpc": "2.0",
    "id": 1,
    "result": "aC4t8i1td8+JWL31qNaV9i1rVc6gRwdPspRJhGTz4CL92MI7by9K/gvkwrtBc7BazNw3Yyu6fuDWs1JUGvEKjA=="
}
```

### CODE REVIEW

File: enclave/src/main/java/org/hyperledger/besu/enclave/types/StoreRawResponse.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.enclave.types;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+
+@JsonPropertyOrder({"key"})
```
Comment (pinges) [inline enclave/src/main/java/org/hyperledger/besu/enclave/types/StoreRawResponse.java:21]: we should not need an order ...

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/methods/PrivJsonRpcMethods.java [hunk 2]
```diff
@@ -106,4 +104,18 @@ protected RpcApi getApiGroup() {
     }
     return RPC_METHODS;
   }
+
+  private JsonRpcMethod distributeRawTransactionMethod(
+      final PrivacyController privacyController,
+      final EnclavePublicKeyProvider enclavePublicKeyProvider) {
+    if (getPrivacyParameters().getGoQuorumPrivacyParameters().isPresent()) {
```
Comment (pinges) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/methods/PrivJsonRpcMethods.java:111]: That should be in the GoQuorumJsonRpsPrivacyMethods

File: enclave/src/main/java/org/hyperledger/besu/enclave/types/StoreRawResponse.java (hunk 1 above)
Comment (macfarla) [reply enclave/src/main/java/org/hyperledger/besu/enclave/types/StoreRawResponse.java:21]: done

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/methods/PrivJsonRpcMethods.java (hunk 2 above)
Comment (macfarla) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/methods/PrivJsonRpcMethods.java]: I changed this but now you get unexpected behaviour if you have PRIV and GOQUORUM APIs enabled

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/methods/PrivacyApiGroupJsonRpcMethods.java [hunk 3]
```diff
@@ -150,12 +150,16 @@ PrivacyQueries getPrivacyQueries() {
 
   private JsonRpcMethod createPrivacyMethod(
       final PrivacyParameters privacyParameters, final JsonRpcMethod rpcMethod) {
-    if (rpcMethod.getName().equals(RpcMethod.ETH_SEND_RAW_PRIVATE_TRANSACTION.getMethodName())) {
+    final String methodName = rpcMethod.getName();
+    if (methodName.equals(RpcMethod.ETH_SEND_RAW_PRIVATE_TRANSACTION.getMethodName())) {
+      return rpcMethod;
+    } else if (methodName.equals(RpcMethod.PRIV_DISTRIBUTE_RAW_TRANSACTION.getMethodName())
```
Comment (pinges) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/methods/PrivacyApiGroupJsonRpcMethods.java:156]: can this be removed now?

**Review by pinges — APPROVED** [review]
I've added another comment!
Would be nice if we could make sure that the right distribute RPC is used when both, PRIV and GOQUORUM RPCs are enabled!

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/methods/PrivacyApiGroupJsonRpcMethods.java (hunk 3 above)
Comment (macfarla) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/methods/PrivacyApiGroupJsonRpcMethods.java]: not quite

**Comment by macfarla** [comment]
> I've added another comment!
> Would be nice if we could make sure that the right distribute RPC is used when both, PRIV and GOQUORUM RPCs are enabled!

to achieve this I have given them different method names

**Pull request merged**